### PR TITLE
Include layout when rendering objects from controllers

### PR DESCRIPTION
**Pull request by joelhawksley**
In https://github.com/rails/rails/pull/37919, support for rendering objects that respond to `render_in` in controllers was added. However, the implementation did not support layouts.

This change updates the implementation from #37919 to be consistent with the ActionView::Template classes, enabling the use of layouts.

Also,  in the interest of making it clear that ViewComponent is not part of Rails, I've renamed all uses of the term
"component" to "renderable".

Co-authored-by: Felipe Sateler <fsateler@gmail.com>

cc @tenderlove

### CODE REVIEW

**Review by eileencodes — CHANGES_REQUESTED** [review]
There's some style issues here that I've made suggested changes for. Be sure to fix those up into a single commit.

I'm not sure `Component` is the right name for this class because it might be confusing with `ViewComponent`.

**Comment by joelhawksley** [comment]
> I'm not sure `Component` is the right name for this class because it might be confusing with `ViewComponent`.

Good point. I went with `RenderInObject`, but perhaps others could think of a better name? We also considered:

```
Renderable
IsolatedRenderable
RenderInable
```

File: actionview/lib/action_view/template/render_in_object.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module ActionView # :nodoc:
```
Comment (eileencodes) [inline actionview/lib/action_view/template/render_in_object.rb:3]:
```suggestion
module ActionView
```

File: actionview/lib/action_view/template/render_in_object.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  class Template # :nodoc:
```
Comment (eileencodes) [inline actionview/lib/action_view/template/render_in_object.rb:4]: Only the new class should be no-doc'd.

```suggestion
  class Template
```

File: actionview/lib/action_view/template/render_in_object.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      def render(context, *args)
+        @component.render_in(context)
+      end
+
+      def format
+        @component.format
+      end
+
+      def formats; Array(format); end
+      deprecate :formats
```
Comment (eileencodes) [inline actionview/lib/action_view/template/render_in_object.rb:27]: I see that this file is very similar to html and text. Maybe this needs a better name and a parent class to inherit from so we can reduce some of the duplication.

This looked to me like a brand new class that was deprecating a brand new method. I think there is some room for an improvement to the template object.

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: actionview/lib/action_view/template/render_in_object.rb (hunk 3 above)
Comment (joelhawksley) [reply actionview/lib/action_view/template/render_in_object.rb]: @eileencodes makes sense. Perhaps we should fix the other ActionView template objects, then? They have the same :nodoc: pattern I used in this PR. I'd be happy to write the PR ❤️

**Comment by joelhawksley** [comment]
@eileencodes thank you for the thoughtful feedback ❤️ 

In responding to your suggestions, the name `Renderable` came to mind. I think it fits this case rather nicely.

I revisited the proposed template object and found that I was able to reduce the interface significantly, to the point where I don't feel there's enough in common with `HTML` and `Text` to warrant extracting a parent class.

**Comment by joelhawksley** [comment]
@eileencodes in the interest of making it clear that ViewComponent is not part of Rails, I've renamed all uses of the term
"component" to "renderable". 

I included this as a separate commit, but let me know if you'd prefer I squash it into the other commit ❤️

**Comment by eileencodes** [comment]
Can you squash this into one commit please?

**Comment by joelhawksley** [comment]
@eileencodes squashed!

**Pull request merged**